src/bsoncxx/types/value.cpp

```diff
@@ -56,35 +74,52 @@ value& value::operator=(const value& rhs) {
 
 value::value(value&& rhs) noexcept {
     *this = std::move(rhs);
+    rhs._type = stdx::nullopt;
 }
 
 value& value::operator=(value&& rhs) noexcept {
-    _type = rhs._type;
+    if(_type && !rhs._type) {
```

Comment:
With the caveat that we should think carefully about self move assign here. @hanumantmk, thoughts?